### PR TITLE
Implement error::Error on all error types

### PR DESCRIPTION
**Pull request by kamalmarhubi**
This lets the `try!` macro work when using `Box<Error>` as a return
type.

### CODE REVIEW

File: rust-manifest/src/errors.rs [hunk 1]
```diff
@@ -13,6 +14,25 @@ pub enum Error {
...
+            ExpectedType(_, _) => "expected type",
+            PackageNotFound(_) => "package not found",
+            TargetNotFound(_) => "target not found",
+            MissingRoot => "manifest has no root package",
+            UnsupportedVersion(_) => "unsupported manifest version",
+        }
+    }
+
+    fn cause(&self) -> Option<&error::Error> {
+        None
```
Comment (kamalmarhubi) [inline rust-manifest/src/errors.rs]: there's an opportunity here to return the first parse error for the `Parsing` variant. easily done if you think that makes sense.

**Comment by Diggsey** [comment]
Why not use the existing `Display` implementation for the error description?

**Comment by kamalmarhubi** [comment]
Lifetimes: return type of `Error::description()` is `&str` with the same lifetime as the error; using the `Display::fmt()` implementation necessitates creating a string which can't live long enough. This made me quite sad.

**Comment by Diggsey** [comment]
I'm still not sure that this is the best way to do this - what are the downsides to creating a wrapper type which uses the display implementation on construction, eg.
```rust
struct BoxedError<E> {
    inner: E,
    message: String,
}
```

As long as the required `From` implementations exist, it should still be fairly seamless?

**Comment by kamalmarhubi** [comment]
@Diggsey I'll give it a shot!

**Comment by kamalmarhubi** [comment]
Thinking about this, I still think the underlying error types should implement `std::error::Error`. 
The cause is lost with this wrapping, since we don't have an `Error` bound on `E`.

As an example, `std` handles it as I did here. Eg, `ffi::NulError`: https://github.com/rust-lang/rust/blob/9cfa1916fa06fa5fa809d6d7807a6f978e35c67d/src/libstd/ffi/c_str.rs#L348-L354

I'm still going to give the suggestion a go though, and see if it works in my use case.

**Comment by kamalmarhubi** [comment]
Take a look at #65 for a take on the wrapper approach.

**Comment by Diggsey** [comment]
@brson Would be glad to hear your thoughts on this!

**Comment by brson** [comment]
I think the approach in this patch - duplicating error messages for every variant - is unfortunately the idiomatic way to do it, so +1.

File: src/errors.rs [hunk 2]
```diff
@@ -118,6 +119,35 @@ impl<'a> Display for Notification<'a> {
...
+        use self::Error::*;
+        match *self {
+            Install(ref e) => error::Error::description(e),
+            Utils(ref e) => error::Error::description(e),
+            Temp(ref e) => error::Error::description(e),
+            UnknownMetadataVersion(_) => "unknown metadata version",
+            InvalidEnvironment => "invalid environment",
+            NoDefaultToolchain => "no default toolchain configured",
+            PermissionDenied => "permission denied",
+            ToolchainNotInstalled(_) => "toolchain '{}' is not installed",
```
Comment (bluss) [inline src/errors.rs]: there's a stray {} in the string here.

Comment (kamalmarhubi) [inline src/errors.rs]: ah nice catch. Fixed.

**Comment by kamalmarhubi** [comment]
@Diggsey let me know if you want me to make this a PR against master rather than new. Also I'm happy to rebase once it's got the go ahead.

**Comment by kamalmarhubi** [comment]
I'm confused by the build error. It is testing against a merge commit that is not included in this PR, and that I did not create.

**Comment by Diggsey** [comment]
I have no idea, looks like a github bug, given that it's happening to both travis and appveyor - try merging from master so that it's completely up to date.

**Comment by kamalmarhubi** [comment]
Just rebased onto new and force pushed. I also tweaked a couple of descriptions. I also got rid of `_` matches so that new variants result in exhaustiveness errors so that proper `cause()` and `description()` implementations get added.

**Comment by kamalmarhubi** [comment]
Actually this needs another few minutes, so don't review just yet.

**Comment by kamalmarhubi** [comment]
ok, I think this is good to review now

**Comment by Diggsey** [comment]
Great, thanks!

**Pull request merged**